kernel/kernel-api/src/test/scala/io/delta/kernel/internal/util/SchemaUtilsSuite.scala

```diff
@@ -895,6 +744,123 @@ class SchemaUtilsSuite extends AnyFunSuite {
+          schemaBefore,
+          schemaAfter,
+          metadata(schemaBefore, tableProperties))
+      }
+
+      assert(e.getMessage.matches("Field .* is missing expected nested column IDs"))
+    }
+  }
+
+  private val mapWithStructKeyInconsistentNestedIds = mapWithStructKey(mapFieldMetadata =
+    FieldMetadata.builder()
+      .fromMetadata(fieldMetadata(id = 1, physicalName = "map"))
+      .putFieldMetadata(
+        COLUMN_MAPPING_NESTED_IDS_KEY,
+        FieldMetadata.builder().putLong("map.key", 8L)
+          .putLong("map.value", 9L).build()).build())
+
+  private val inconsistentNestedIdsEvolution = Table(
+    ("schemaBefore", "updatedSchemaWithNoNestedId"),
+    // Map with struct key with inconsistent nested IDs
+    (
+      mapWithStructKeyIcebergCompatV2,
+      mapWithStructKeyInconsistentNestedIds))
+
+  test("validateUpdatedSchema fails when nestedIds are not consistent in IcebergCompatV2") {
+    forAll(inconsistentNestedIdsEvolution) { (schemaBefore, schemaAfter) =>
+      val e = intercept[IllegalArgumentException] {
+        val tableProperties =
+          Map(COLUMN_MAPPING_MODE_KEY -> "id", "delta.enableIcebergCompatV2" -> "true")
+        SchemaUtils.validateUpdatedSchema(
+          schemaBefore,
+          schemaAfter,
+          metadata(schemaBefore, tableProperties))
+      }
+
+      assert(e.getMessage.matches("Expected field with id .* to have nested key.*"))
+    }
+  }
+
+  test("validateUpdatedSchema with valid nested IDs succeeds in IcebergCompatV2") {
+    val validNestedIdsEvolution =
+      missingNestedIdsEvolution.map(evolution => (evolution._2, evolution._1))
+    forAll(validNestedIdsEvolution) { (schemaBefore, schemaAfter) =>
+      val tableProperties =
+        Map(COLUMN_MAPPING_MODE_KEY -> "id", "delta.enableIcebergCompatV2" -> "true")
+      SchemaUtils.validateUpdatedSchema(
+        schemaBefore,
+        schemaAfter,
+        metadata(schemaBefore, tableProperties))
+    }
+  }
+
+  private def mapWithStructKey(
+      mapType: DataType =
+        mapWithStructKey.get("map").getDataType,
+      mapFieldMetadata: FieldMetadata =
+        mapWithStructKey.get("map").getMetadata): StructType = {
+    new StructType()
+      .add(
+        "map",
+        mapType,
+        true,
+        mapFieldMetadata)
+  }
+
+  private def structWithArrayOfStructs(
+      arrayType: ArrayType =
+        structWithArrayOfStructs
+          .get("top_level_struct")
+          .getDataType.asInstanceOf[StructType]
+          .get("array").getDataType.asInstanceOf[ArrayType],
+      arrayMetadata: FieldMetadata =
+        structWithArrayOfStructs
+          .get("top_level_struct")
+          .getDataType.asInstanceOf[StructType]
+          .get("array").getMetadata) = {
+    new StructType()
+      .add(
+        "top_level_struct",
+        new StructType().add(
+          "array",
+          arrayType,
+          false,
+          arrayMetadata),
+        fieldMetadata(1L, "top_level_struct"))
```

Comment:
Added some helper functions to help reduce duplication, there's very specific inner fields we're modifying for each of these cases so we can just pass in the inner types/inner metadata specific for each case